aip/auth/4120.md

```diff
@@ -0,0 +1,96 @@
+
+This value can also be used to make decisions about whether the library is in the GDU or not e.g. the purpose of throwing specific errors if the end-user is forcing an unsupported authentication flow with non-GDU credentials.
+
+#### Authentication libraries **must** use the self-signed JWT flow when authenticating outside the GDU using Service Account credentials
+
+Currently the self-signed JWT (SSJ) sub-flow is the default option for Service Account auth flow, but user parameters provided, such as `scope`, `audience`, or `useJwtWithScopes` can change that. E.g. setting  `useJwtWithScopes` to `false` when specifying `scope` would result in auth library fallback to token exchange flow for service accounts.
+The token exchange for Service Accounts is not a supported flow outside the GDU, and services must support SSJ with scopes. Therefore authentication libraries **must** use the self-signed JWT flow when authenticating outside the GDU using Service Account credentials. If the `useJwtWithScopes` parameter exists (whatever form it takes in specific languages), it **must** be ignored outside the GDU, and its documentation **must** be updated to reflect that.
+
+### GCE credentials retrieve the universe domain from the metadata server
+
+> This requirement applies **only** to the VM metadata flow.
+
+GCE credentials (also known as VM Metadata credentials) directly retrieves an access token from the Metadata Server (MDS). The access token has no universe information. The authentication libraries **must** modify the VM metadata flow to make a second request to retrieve the universe_domain from the following URI: `http://metadata.google.internal/computeMetadata/v1/universe/universe_domain`.
+
+Once the `universe_domain` is retrieved, the authentication libraries **must** cache it to avoid making unnecessary MDS requests.
+
+Failure modes when retrieving the universe_domain from MDS:
+
+| Error when retrieving the universe_domain from MDS | Action |
+| -------------------------------------------------- | ------ |
+| 404 | The authentication libraries **must** set the `universe_domain` to the default value `"googleapis.com"`. |
+| Timeout | The authentication libraries **must not** fall back to the `"googleapis.com"` default value. A retry mechanism is recommended. The authentication libraries **should** do the same action as when retrieving the access token times out. |
+| Any other error | The authentication libraries **must not** fall back to the `"googleapis.com"` default value. The authentication libraries **should** do the same action as when the same error happens when retrieving the access token. |
```

Comment:
Heads up: this has turned out to be a bit of a problem -- because "MDS" doesn't much have a specification we can refer to, customers have implemented "MDS" clones that don't match our expectations.

Beyond the scope of this AIP, but is there a way we can mitigate that problem without requiring changes to MDS clones or our libraries?  Maybe a  trusted parameter or environment variable available to auth libraries to fix the universe domain.

Reply:
By accepting 404s to be `googleapis.com`, we should be handling all edgecases where other MDSs don't behave as expected (see [line 66](https://github.com/aip-dev/google.aip.dev/pull/1282/files/665af2ce7f5291612b2485b45f410bb7348ecddc#diff-79bdafd58065e6a283cb549b61352d41286e5f26635598885c6c1f10b3d5e76cR66))

There's a potential for accepting a timeout as `googleapis.com` in the event that the library has _already_ confirmed that it's running on MDS. But this is a temporary fix, and has not been accepted in the spec yet, so I think it's best to leave it out.